packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -400,12 +477,47 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
       TurboModulePerfLogger::asyncMethodCallExecutionStart(moduleName, methodNameStr.c_str(), asyncCallCounter);
     }
 
-    @try {
+    try {
       [inv invokeWithTarget:strongModule];
-    } @catch (NSException *exception) {
-      throw convertNSExceptionToJSError(runtime, exception);
-    } @finally {
-      [retainedObjectsForInvocation removeAllObjects];
+    } catch (NSException *exception) {
+      NSDictionary *cause = @{
+        @"name": exception.name,
+        @"message": exception.reason,
+        @"stackSymbols": exception.callStackSymbols,
+        @"stackReturnAddresses": exception.callStackReturnAddresses,
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (NSError *error) {
+      NSDictionary *cause = @{
+        @"userInfo": error.userInfo,
+        @"domain": error.domain,
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (NSString *errorMessage) {
+      NSDictionary *cause = @{
+        @"message": errorMessage,
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (id e) {
+      NSDictionary *cause = @{
+        @"message": @"Unknown Objective-C Object thrown.",
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (const std::exception &exception) {
+      NSDictionary *cause = @{
+        @"message": [NSString stringWithUTF8String:exception.what()],
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (const std::string &errorMessage) {
+      NSDictionary *cause = @{
+        @"message": [NSString stringWithUTF8String:errorMessage.c_str()],
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
+    } catch (...) {
+      NSDictionary *cause = @{
+        @"message": @"Unknown C++ exception thrown.",
+      };
+      handleCause(runtime, isSync, cause, retainedObjectsForInvocation, optionalInternalRejectBlock);
     }
```

Comment:
nit: inline the cause dictionary in the method call below, and keep the `finally` logic here to be explicit.